wallet/src/account/mod.rs

```diff
@@ -0,0 +1,407 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key_chain::AccountKeyChain;
+use crate::send_request::address_output;
+use crate::{SendRequest, WalletError, WalletResult};
+use common::address::Address;
+use common::chain::signature::inputsig::standard_signature::StandardInputSignature;
+use common::chain::signature::inputsig::InputWitness;
+use common::chain::signature::sighash::sighashtype::SigHashType;
+use common::chain::signature::TransactionSigError;
+use common::chain::tokens::{OutputValue, TokenData, TokenId};
+use common::chain::{
+    Block, ChainConfig, Destination, OutPoint, OutPointSourceId, SignedTransaction, TxOutput,
+};
+use common::primitives::{Amount, BlockHeight, Idable};
+use crypto::key::extended::ExtendedPrivateKey;
+use crypto::key::hdkd::u31::U31;
+use std::collections::BTreeMap;
+use std::ops::Add;
+use std::sync::Arc;
+use storage::Backend;
+use utxo::{Utxo, UtxoSource};
+use wallet_storage::{StoreTxRo, StoreTxRw, WalletStorageRead, WalletStorageWrite};
+use wallet_types::{AccountId, AccountOutPointId, KeyPurpose};
+
+pub struct Account {
+    chain_config: Arc<ChainConfig>,
+    key_chain: AccountKeyChain,
+}
+
+impl Account {
+    pub fn load_from_database<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        db_tx: &StoreTxRo<B>,
+        id: &AccountId,
+        root_key: &ExtendedPrivateKey,
+    ) -> WalletResult<Account> {
+        let key_chain =
+            AccountKeyChain::load_from_database(chain_config.clone(), db_tx, id, root_key)?;
+
+        Ok(Account {
+            chain_config,
+            key_chain,
+        })
+    }
+
+    /// Create a new account by providing a key chain
+    pub fn new<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        db_tx: &mut StoreTxRw<B>,
+        key_chain: AccountKeyChain,
+    ) -> WalletResult<Account> {
+        let account_id = key_chain.get_account_id();
+        let account_info = key_chain.get_account_info();
+
+        db_tx.set_account(&account_id, &account_info)?;
+
+        let mut account = Account {
+            chain_config,
+            key_chain,
+        };
+
+        account.scan_genesis_block(db_tx)?;
+
+        Ok(account)
+    }
+
+    pub fn process_send_request<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        mut request: SendRequest,
+    ) -> WalletResult<SignedTransaction> {
+        if request.utxos().is_empty() {
+            let utxos: BTreeMap<OutPoint, Utxo> = db_tx
+                .get_utxo_set(&self.get_account_id())?
+                .into_iter()
+                .map(|(outpoint, utxo)| (outpoint.into_item_id(), utxo))
+                .collect();
+
+            // TODO: Call coin selector
+
+            request.fill_inputs(utxos)?;
+        }
+
+        let (input_coin_amount, input_tokens_amounts) =
+            Self::calculate_output_amounts(request.utxos().iter())?;
```

Comment:
Just thinking out loud: should we consider in the future reusing this functionality from TransactionVerifier? It could be beneficial in 2 ways: avoid code duplication and unifying interfaces because it will require implementing `UtxoView` and `PoSAccountingView`. `UtxoView` and `UtxoCache` are good abstractions to replace the current approach with BtreeMap

Reply:
I'm not sure about this idea. I would prefer to keep wallet's code separately.

Reply:
To be fair, there's tons of code that's implemented in the TxVerifier and Gosha has a fair point. You're just starting with the wallet and UTXOs, Pavel, and you'll have to reinvent everything. Perhaps it's a good idea to try to reuse code when we can. Again, as usual, not right now for this particular problem, but in general we want to avoid code duplication.